### PR TITLE
[63_2] Merge contiguous indices for sorted bib items

### LINKED ISSUE #166: Sort and merge citations upon exporting to PDF
[Sort and merge citations upon exporting to PDF](https://github.com/texmacs/texmacs/pull/24)

### PR DESCRIPTION
**Pull request by jingkaimori**
## Why you open this Pull Request?

Merge continuous citing indices, close #166 :
| Origin | Merged |
|--------|--------|
| [1, 2, 3] | [1-3] |
| [1, 2, 3, 4, 5] | [1-5] |
| [1, 3, 4, 5, 7] | [1, 3-5, 7] |
| [1, 2, 4, 5] | [1, 2, 4, 5] | 

Users should add `cite-sort` style package explicitly to enable this feature.

[Origin patch is provided by Zhengfei Hu(hammer)](https://savannah.gnu.org/patch/?10339) @hammerfunctor 

## What work have you done in the current Pull Request?

- [x] import Zhengfei Hu's work
- [x] improve origin patch slightly

### CODE REVIEW

File: TeXmacs/progs/utils/cite/cite-sort.scm [hunk 1]
```diff
@@ -18,18 +18,39 @@
...
      (s1 (if (string? t1) t1 (convert t1 "stm-snippet" "verbatim-snippet")))
      (s2 (if (string? t2) t2 (convert t2 "stm-snippet" "verbatim-snippet"))))
-    (compare-string s1 s2)))
+    (comparator s1 s2)))
 
 (define (expand-references k)
   (with key (stree->tree `(get-binding ,(cadr k)))
     (with ret (tree->stree (texmacs-exec key))
       (if (!= ret '(uninit)) ret ""))))
+(define (merge-contiguous new old present)
```
Comment (darcy-shen) [inline TeXmacs/progs/utils/cite/cite-sort.scm:38]: extra line please

File: TeXmacs/progs/utils/cite/cite-sort.scm [hunk 2]
```diff
@@ -18,18 +18,39 @@
...
+            new
+            (append new (flush)))
+        (if (null? present)
+            (merge-contiguous new (cdr old) (list (car old)))
+            (if (compare-cite-keys (cAr present) (car old) compare-string-inc?)
+                (merge-contiguous new (cdr old) (append present (list (car old))))
+                (merge-contiguous (append new (flush)) (cdr old) (list (car old)))
+                )))))
 
 (tm-define (cite-sort args)
```
Comment (darcy-shen) [inline TeXmacs/progs/utils/cite/cite-sort.scm:55]: unit test on the affected interface

**Comment by darcy-shen** [comment]
The develop workflow is very smooth now on macOS and Linux: https://github.com/XmacsLabs/mogan/blob/branch-1.2/docs/guide/Develop_on_macOS.md

You may try it if you have time.

**Comment by darcy-shen** [comment]
image


Tried, it failed to work for me. I rebased this branch to branch-1.2 and tested.

**Comment by darcy-shen** [comment]
works fine now with cite-sort package attached.

File: TeXmacs/progs/utils/cite/cite-sort.scm [hunk 3]
```diff
@@ -18,18 +18,39 @@
...
+    (if (null? old)
+        (if (null? present)
+            new
+            (append new (flush)))
+        (if (null? present)
+            (merge-contiguous new (cdr old) (list (car old)))
+            (if (compare-cite-keys (cAr present) (car old) compare-string-inc?)
+                (merge-contiguous new (cdr old) (append present (list (car old))))
+                (merge-contiguous (append new (flush)) (cdr old) (list (car old)))
+                )))))
```
Comment (darcy-shen) [inline TeXmacs/progs/utils/cite/cite-sort.scm:53]: scheme code format

File: TeXmacs/progs/utils/cite/cite-sort-test.scm [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+     '(("1" (reference "bib1"))
+       ("3" (reference "bib3"))
+       ("4" (reference "bib4"))
+       ("2" (reference "bib2"))
+       ("5" (reference "bib5")))
+     '((concat
+         (reference "bib1")
+         "-"
+         (reference "bib5"))))
+))
```
Comment (darcy-shen) [inline TeXmacs/progs/utils/cite/cite-sort-test.scm:65]: this is ok, because there are test groups above

**Review by darcy-shen — APPROVED** [review]
Awesome work!

**Comment by hammerfunctor** [comment]
When this is turned on, references disappear every second update, where links to those hidden references are not present and thereby corresponding references are considered not cited at all. One more update after that brings everything back.

**Comment by darcy-shen** [comment]
>  One more update after that brings everything back.
Is it a fix or just a revert?

**Pull request merged**